lib/gcloud/storage/bucket/list.rb

```diff
@@ -51,6 +51,67 @@ def next
         end
 
         ##
+        # Retrieves all buckets by repeatedly loading {#next} until {#next?}
+        # returns `false`. Calls the given block once for each bucket, which is
+        # passed as the parameter.
+        #
+        # An Enumerator is returned if no block is given.
+        #
+        # This method may make several API calls until all buckets are
+        # retrieved. Be sure to use as narrow a search criteria as possible.
+        # Please use with caution.
+        #
+        # @example Iterating each bucket by passing a block:
+        #   require "gcloud"
+        #
+        #   gcloud = Gcloud.new
+        #   storage = gcloud.storage
+        #
+        #   buckets = storage.buckets
+        #   buckets.all do |bucket|
+        #     puts bucket.name
+        #   end
+        #
+        # @example Using the enumerator by not passing a block:
+        #   require "gcloud"
+        #
+        #   gcloud = Gcloud.new
+        #   storage = gcloud.storage
+        #
+        #   buckets = storage.buckets
+        #   all_names = buckets.all.map do |bucket|
+        #     bucket.name
+        #   end
+        #
+        # @example Limit the number of API calls made:
+        #   require "gcloud"
+        #
+        #   gcloud = Gcloud.new
+        #   storage = gcloud.storage
+        #
+        #   buckets = storage.buckets
+        #   buckets.all(max_api_calls: 10) do |bucket|
+        #     puts bucket.name
+        #   end
+        #
+        def all max_api_calls: nil
+          max_api_calls = max_api_calls.to_i if max_api_calls
+          unless block_given?
+            return enum_for(:all, max_api_calls: max_api_calls)
+          end
+          results = self
+          loop do
+            results.each { |r| yield r }
+            if max_api_calls
+              max_api_calls -= 1
+              break if max_api_calls < 0
```

Comment:
Because this `break` is after the `yield`, this corner case is possible:

```ruby
irb(main):015:0> storage.buckets.all(max_api_calls: 0).each {|x| puts x.name};nil
my-bucket123321
=> nil
```